xcm/pallet-xcm/src/lib.rs

```diff
@@ -448,9 +597,203 @@ pub mod pallet {
+		///
+		/// Errors:
+		/// - `BadLocation`: If the `location` cannot be expressed with the XCM version used by this
+		///   chain.
+		#[pallet::weight(100_000_000u64)]
+		pub fn force_xcm_version(
+			origin: OriginFor<T>,
+			location: Box<MultiLocation>,
+			xcm_version: XcmVersion,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			let location = *location;
+			SupportedVersion::<T>::insert(
+				XCM_VERSION,
+				LatestVersionedMultiLocation(&location),
+				xcm_version,
+			);
+			Self::deposit_event(Event::SupportedVersionChanged(location, xcm_version));
+			Ok(())
+		}
+
+		/// Set a safe XCM version (the version that XCM should be encoded with if the most recent
+		/// version a destination can accept is unknown).
+		///
+		/// - `origin`: Must be Root.
+		/// - `maybe_xcm_version`: The default XCM encoding version, or `None` to disable.
+		#[pallet::weight(100_000_000u64)]
+		pub fn force_default_xcm_version(
+			origin: OriginFor<T>,
+			maybe_xcm_version: Option<XcmVersion>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			SafeXcmVersion::<T>::set(maybe_xcm_version);
+			Ok(())
+		}
+
+		/// Ask a location to notify us regarding their XCM version and any changes to it.
+		///
+		/// - `origin`: Must be Root.
+		/// - `location`: The location to which we should subscribe for XCM version notifications.
+		#[pallet::weight(100_000_000u64)]
+		pub fn force_subscribe_version_notify(
+			origin: OriginFor<T>,
+			location: Box<VersionedMultiLocation>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			let location = (*location).try_into().map_err(|()| Error::<T>::BadLocation)?;
+			Self::request_version_notify(location).map_err(|e| {
+				match e {
+					XcmError::InvalidLocation => Error::<T>::AlreadySubscribed,
+					_ => Error::<T>::InvalidOrigin,
+				}
+				.into()
+			})
+		}
+
+		/// Require that a particular destination should no longer notify us regarding any XCM
+		/// version changes.
+		///
+		/// - `origin`: Must be Root.
+		/// - `location`: The location to which we are currently subscribed for XCM version
+		///   notifications which we no longer desire.
+		#[pallet::weight(100_000_000u64)]
+		pub fn force_unsubscribe_version_notify(
+			origin: OriginFor<T>,
+			location: Box<VersionedMultiLocation>,
+		) -> DispatchResult {
+			ensure_root(origin)?;
+			let location = (*location).try_into().map_err(|()| Error::<T>::BadLocation)?;
+			Self::unrequest_version_notify(location).map_err(|e| {
+				match e {
+					XcmError::InvalidLocation => Error::<T>::NoSubscription,
+					_ => Error::<T>::InvalidOrigin,
+				}
+				.into()
+			})
+		}
 	}
 
 	impl<T: Config> Pallet<T> {
+		fn check_xcm_version_change() -> Weight {
+			// We assume that supported XCM version only ever increases, so just cycle through lower
+			// XCM versioned from the current.
+			for v in 0..XCM_VERSION {
+				for (old_key, value) in SupportedVersion::<T>::drain_prefix(v) {
```

Comment:
it is important that this will be bounded, which i believe it is to the number of parachains in the immediate relay chain system, but may not scale to the nested relay chain future with unbounded number of locations.

Reply:
probably same below

Reply:
i saw later in the review that you dont allow sub-locations, which may resolve this point. however, i think it would be better to design this logic only taking N items at a time, where N is some defined upper limit, and handling not being able to go through all the items at once.

Reply:
done.